model/user/user.go

```diff
@@ -166,6 +168,53 @@ func (u *DBUser) UpdateParsleySettings(settings parsley.Settings) error {
 	return nil
 }
 
+// CheckAndUpdateSchedulingLimit checks if the number of tasks to be activated by the user is allowed given
+// the global per-user hourly task scheduling limit, and updates relevant timestamp and counter info used
+// to track the user's hourly scheduling usage. The numTasksActivated parameter being negative signifies
+// the user is deactivating tasks, which frees up space in their scheduling limit.
+func (u *DBUser) CheckAndUpdateSchedulingLimit(settings *evergreen.Settings, numTasksActivated int) error {
+	var update bson.M
+	maxScheduledTasks := settings.TaskLimits.MaxHourlyPatchTasks
+	if maxScheduledTasks == 0 {
+		return nil
+	}
+	if numTasksActivated > maxScheduledTasks {
+		return errors.Errorf("cannot schedule %d tasks, maxumum hourly per-user limit is %d", numTasksActivated, maxScheduledTasks)
+	}
+	now := time.Now()
+	oneHourAgo := now.Add(-1 * time.Hour)
+	// If the last time the user scheduled patch tasks was within the hour, increment the number
+	// of activated tasks to the user's counter, erroring if the global limit is breached. If numTasksActivated
+	// is negative, the counter is decremented.
+	if u.LastScheduledTasksAt.After(oneHourAgo) {
+		update = bson.M{
+			"$set": bson.M{NumScheduledPatchTasksKey: getNewCounter(u.NumScheduledPatchTasks, numTasksActivated)},
+		}
+		if (numTasksActivated + u.NumScheduledPatchTasks) >= maxScheduledTasks {
+			minutesRemaining := 60 - int(now.Sub(u.LastScheduledTasksAt).Minutes())
+			return errors.Errorf("user '%s' has scheduled %d out of %d allowed tasks in the past hour, limit refreshes in %d minutes", u.Id, u.NumScheduledPatchTasks, maxScheduledTasks, minutesRemaining)
+		}
+	} else {
+		// Otherwise, if the user has not scheduled any patch tasks within the past hour, reset the last scheduled tasks
+		// timestamp to now, and reset the number of schedule tasks to the number of activated tasks passed in here. If we are dea
+		update = bson.M{
+			"$set": bson.M{
+				NumScheduledPatchTasksKey: getNewCounter(0, numTasksActivated),
+				LastScheduledTasksAtKey:   time.Now(),
+			},
+		}
+	}
+	return UpdateOne(bson.M{IdKey: u.Id}, update)
+}
+
+func getNewCounter(currentCounter, numTasksActivated int) int {
+	sum := currentCounter + numTasksActivated
+	if sum < 0 {
```

Comment:
can we also set this to return the limit as the max number because we might fail validation if a user tries to deactivate tasks if their counter is super high

Reply:
maybe we can add that as a test case?

Reply:
Can you clarify what you mean? If a user deactivates tasks then their current counter minus how many tasks they're deactivating will be evaluated against the limit [here](https://github.com/evergreen-ci/evergreen/pull/7765/files#diff-fd21a2068e029268ed8633e2d8e5752b619a35b3525ef5e25c2686a4243c48c5R193)

Reply:
I misunderstood, I thought we were setting NumScheduledPatchTasks to `getNewCounter(u.NumScheduledPatchTasks, numTasksModified)},` even if we go above the limit. But that isn't the case. Resolving